pkg/util/informermanager/informermanager.go

```diff
@@ -0,0 +1,287 @@
+
+		m.informers[ftcName] = informer
+		m.informerStopChs[ftcName] = stopCh
+		m.eventHandlerRegistrations[ftcName] = map[*EventHandlerGenerator]cache.ResourceEventHandlerRegistration{}
+		m.lastAppliedFTCsCache[ftcName] = map[*EventHandlerGenerator]*fedcorev1a1.FederatedTypeConfig{}
+	}
+
+	registrations := m.eventHandlerRegistrations[ftcName]
+	lastAppliedFTCs := m.lastAppliedFTCsCache[ftcName]
+
+	ftc = ftc.DeepCopy()
+
+	for _, generator := range m.eventHandlerGenerators {
+		lastApplied := lastAppliedFTCs[generator]
+		if !generator.Predicate(lastApplied, ftc) {
+			lastAppliedFTCs[generator] = ftc
+			continue
+		}
+
+		if oldRegistration := registrations[generator]; oldRegistration != nil {
+			if err := informer.Informer().RemoveEventHandler(oldRegistration); err != nil {
+				return fmt.Errorf("failed to unregister event handler: %w", err), true
+			}
+			delete(registrations, generator)
+		}
+		delete(lastAppliedFTCs, generator)
+
+		if handler := generator.Generator(ftc); handler != nil {
+			newRegistration, err := informer.Informer().AddEventHandler(handler)
+			if err != nil {
+				return fmt.Errorf("failed to register event handler: %w", err), true
+			}
+			registrations[generator] = newRegistration
+		}
+
+		lastAppliedFTCs[generator] = ftc
+	}
+
+	return nil, false
+}
+
+func (m *informerManager) processFTCDeletion(ftcName string) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.processFTCDeletionUnlocked(ftcName)
+}
+
+func (m *informerManager) processFTCDeletionUnlocked(ftcName string) error {
+	if stopCh, ok := m.informerStopChs[ftcName]; ok {
+		close(stopCh)
+	}
+
+	m.gvrMapping.Delete(ftcName)
+
+	delete(m.informers, ftcName)
+	delete(m.informerStopChs, ftcName)
+	delete(m.eventHandlerRegistrations, ftcName)
+
+	return nil
+}
+
+func (m *informerManager) AddEventHandlerGenerator(generator *EventHandlerGenerator) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.started {
+		return fmt.Errorf("InformerManager is already started.")
+	}
+
+	m.eventHandlerGenerators = append(m.eventHandlerGenerators, generator)
+	return nil
+}
+
+func (m *informerManager) GetFederatedTypeConfigLister() v1alpha1.FederatedTypeConfigLister {
+	return m.ftcInformer.Lister()
+}
+
+func (m *informerManager) GetResourceLister(
+	gvr schema.GroupVersionResource,
+) (lister cache.GenericLister, informerSynced cache.InformerSynced, exists bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	ftc, ok := m.gvrMapping.ReverseLookup(gvr)
+	if !ok {
+		return nil, nil, false
+	}
+
+	informer, ok := m.informers[ftc]
+	if !ok {
+		return nil, nil, false
+	}
+
+	return informer.Lister(), informer.Informer().HasSynced, true
+}
+
+func (m *informerManager) HasSynced() bool {
+	return m.ftcInformer.Informer().HasSynced()
+}
+
+func (m *informerManager) Start(ctx context.Context) {
+	if !cache.WaitForNamedCacheSync("informer-manager", ctx.Done(), m.HasSynced) {
+		return
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.started {
```

Comment:
Should this be performed before `WaitForNamedCacheSync`? `WaitForNamedCacheSync` prints logs each time it's called.